### PR TITLE
add function applyDiff to source;

### PR DESCRIPTION
**Pull request by sofiia-tesliuk**

### CODE REVIEW

File: interdiff.go [hunk 1]
```diff
@@ -88,9 +88,104 @@ func InterDiffPath(sourcePath string, oldDiff, newDiff io.Reader) (string, error
...
+	for _, hunk := range diffFile.Hunks {
+		// Add untouched part of source
+		newBody = append(newBody, sourceBody[currentOrgSourceI:hunk.OrigStartLine]...)
+		currentOrgSourceI = hunk.OrigStartLine
+
+		hunkBody := strings.Split(string(hunk.Body), "\n")
+		if hunkBody[len(hunkBody)-1] == "" {
+			// Delete last empty line (produced by Split)
+			hunkBody = hunkBody[:len(hunkBody)-1]
+		}
```
Comment (schachmat) [inline interdiff.go:156]:
```suggestion
		hunkBody := strings.Split(strings.TrimSuffix(string(hunk.Body), "\n"), "\n")
```

File: interdiff.go [hunk 2]
```diff
@@ -88,9 +88,104 @@ func InterDiffPath(sourcePath string, oldDiff, newDiff io.Reader) (string, error
...
+						currentOrgSourceI, line[1:], sourceBody[currentOrgSourceI])
+					return "", errors.New("content of original lines in source and diff aren't same")
+				}
+
+				if strings.HasPrefix(line, " ") {
+					newBody = append(newBody, sourceBody[currentOrgSourceI-1])
+				}
+
+				currentOrgSourceI++
+			}
```
Comment (schachmat) [inline interdiff.go:178]:
```suggestion
			if strings.HasPrefix(line, "+") {
				newBody = append(newBody, line[1:])
			} else if strings.HasPrefix(line, " ") {
				newBody = append(newBody, sourceBody[currentOrgSourceI-1])
			} else {
				// (currentOrgSourceI - 1) -- In diff lines started counting from 1
				if line[1:] != sourceBody[currentOrgSourceI-1] {
					fmt.Printf("currentOrgSourceI: %d\nDiff:\t%s\nSource:\t%s\n",
						currentOrgSourceI, line[1:], sourceBody[currentOrgSourceI])
					return "", errors.New("content of original lines in source and diff aren't same")
				}

				currentOrgSourceI++
			}
```

File: interdiff.go [hunk 3]
```diff
@@ -88,9 +88,104 @@ func InterDiffPath(sourcePath string, oldDiff, newDiff io.Reader) (string, error
...
+
+			if strings.HasPrefix(line, "+") {
+				newBody = append(newBody, line[1:])
+			} else {
+				// (currentOrgSourceI - 1) -- In diff lines started counting from 1
+				if line[1:] != sourceBody[currentOrgSourceI-1] {
+					fmt.Printf("currentOrgSourceI: %d\nDiff:\t%s\nSource:\t%s\n",
+						currentOrgSourceI, line[1:], sourceBody[currentOrgSourceI])
+					return "", errors.New("content of original lines in source and diff aren't same")
```
Comment (schachmat) [inline interdiff.go:170]: Use `fmt.Errorf()` to add the lines via `%q` (=quoted, makes it easier to see diff in whitespace) to the error message.

File: interdiff.go [hunk 4]
```diff
@@ -88,9 +88,104 @@ func InterDiffPath(sourcePath string, oldDiff, newDiff io.Reader) (string, error
 	return "", nil
 }
 
+// MixedModeDiff does things
+func MixedMode(oldSource, newSource io.Reader, oldFileDiff, newFileDiff *diff.FileDiff) (string, error) {
```
Comment (schachmat) [inline interdiff.go:92]: Please add testcases for this new func.

File: interdiff.go [hunk 5]
```diff
@@ -88,9 +88,104 @@ func InterDiffPath(sourcePath string, oldDiff, newDiff io.Reader) (string, error
...
+		return "", fmt.Errorf("error applying diff to NewSource %w", errUpdNewSource)
+	}
+
+	// TODO: Get diff from updatedOldSource and updatedNewSource
+
+	// TODO: Convert got diff (chunks) in to FileDiff
+
+	// TODO: Manage timestamps
+
+	// TODO: Get content of FileDiff
```
[the remark this answers is not in the document]
Comment (sofiia-tesliuk) [reply interdiff.go:121]: Planning to implement this part in another PR.

File: interdiff.go (hunk 3 above)
Comment (sofiia-tesliuk) [reply interdiff.go:170]: This line was added for manual debugging. 
Deleted this `Printf`.

File: interdiff.go (hunk 2 above)
Comment (sofiia-tesliuk) [reply interdiff.go:178]: In this case we're checking the content of original line only, when it is deleted. Same for updating currentOrgSourceI value, but this should be done for both cases (for "-" deleted and " " unchanged lines).

File: interdiff.go (hunk 4 above)
Comment (sofiia-tesliuk) [reply interdiff.go:93]: Added testcases for ApplyDiff function, since Mixed isn't finished yet.

File: interdiff_test.go [hunk 6]
```diff
@@ -18,6 +19,20 @@ var interDiffFileTests = []struct {
...
+	diffFile        string
+	resultFile      string
+	finishedWithErr bool
+}{
+	{"test_examples/file_1.txt", "test_examples/diff_1_1_a.txt", "test_examples/file_1_a.txt", false},
+	{"test_examples/file_1.txt", "test_examples/diff_1_1_b.txt", "test_examples/file_1_b.txt", false},
+	{"test_examples/file_2.txt", "test_examples/diff_1_2_a.txt", "test_examples/file_2_a.txt", false},
+	{"test_examples/file_2.txt", "test_examples/diff_1_2_b.txt", "test_examples/file_2_b.txt", false},
+	// sourceFile and diffFile have different origin content.
+	{"test_examples/file_1.txt", "test_examples/diff_1_1_a_wrong_origin.txt", "test_examples/file_1_a.txt", true},
```
Comment (schachmat) [inline interdiff_test.go:33]:
```suggestion
	{"test_examples/file_1.txt", "test_examples/diff_1_a.txt", "test_examples/file_1_a.txt", false},
	{"test_examples/file_1.txt", "test_examples/diff_1_b.txt", "test_examples/file_1_b.txt", false},
	{"test_examples/file_2.txt", "test_examples/diff_2_a.txt", "test_examples/file_2_a.txt", false},
	{"test_examples/file_2.txt", "test_examples/diff_2_b.txt", "test_examples/file_2_b.txt", false},
	// sourceFile and diffFile have different origin content.
	{"test_examples/file_1.txt", "test_examples/diff_1_a_wrong_origin.txt", "test_examples/file_1_a.txt", true},
```

Did I miss something or can we simplify the file names a bit here?

File: interdiff_test.go [hunk 7]
```diff
@@ -18,6 +19,20 @@ var interDiffFileTests = []struct {
 	{"test_examples/result_2.txt", "test_examples/diff_2_a.txt", "test_examples/diff_2_b.txt"},
 }
 
+var applyDiffFileTests = []struct {
+	sourceFile      string
+	diffFile        string
+	resultFile      string
+	finishedWithErr bool
```
Comment (schachmat) [inline interdiff_test.go:26]:
```suggestion
	wantErr bool
```

The convention for test case terminology is to use "got" and "want". When throwing test failures you should also incorporate both, "got" first, then "want". Something like this:

```
t.Errorf("Failed to apply diff %q: got error: %v; want error: nil", tt.diffFile, err)
```

You should also do that for the rest of the test cases, but no need to do it in this PR.

File: interdiff_test.go [hunk 8]
```diff
@@ -68,3 +83,52 @@ func TestInterDiffMode(t *testing.T) {
 		})
 	}
 }
+
+func TestApplyDiff(t *testing.T) {
+	for _, tt := range applyDiffFileTests {
+		t.Run(tt.resultFile, func(t *testing.T) {
```
Comment (schachmat) [inline interdiff_test.go:89]: the following `{ … }` block has many empty lines. Try to use empty lines sparingly. They should separate logical blocks. For example it's very uncommon to have an empty line before the `if err != nil` error handling block or between multiple lines of variable initialization.

File: interdiff_test.go [hunk 9]
```diff
@@ -68,3 +83,52 @@ func TestInterDiffMode(t *testing.T) {
...
+
+			diffFile, err := os.Open(tt.diffFile)
+
+			if err != nil {
+				t.Errorf("Error opening diffFile %q", tt.diffFile)
+			}
+
+			var diffReader io.Reader = diffFile
+
+			diffFileDiff, err := diff.NewFileDiffReader(diffReader).Read()
```
Comment (schachmat) [inline interdiff_test.go:104]:
```suggestion
			d, err := diff.NewFileDiffReader(diffFile).Read()
```

You should be able to use `diffFile` as an `io.Reader` directly.

Also golang folks like short variable names, so maybe try calling this one just `d`.

File: interdiff_test.go [hunk 10]
```diff
@@ -68,3 +83,52 @@ func TestInterDiffMode(t *testing.T) {
...
+			} else {
+				if err != nil {
+					t.Error(err)
+				}
+
+				if !bytes.Equal(normalizeNewlines([]byte(currentResult)), normalizeNewlines(correctResult)) {
+					t.Errorf("File contents mismatch for %s.\nExpected:\n%s\nGot:\n%s\n",
+						tt.resultFile, correctResult, currentResult)
+				}
+			}
```
Comment (schachmat) [inline interdiff_test.go:131]:
```suggestion
			if tt.wantErr && err == nil {
				t.Errorf("Applying diff for %q: got error nil; want error non-nil", tt.resultFile)
			} else if !tt.wantErr {
				if err != nil {
					t.Errorf("Applying diff for %q: got error %v; wan error nil", tt.resultFile, err)
				}
				
				if !bytes.Equal(normalizeNewlines([]byte(currentResult)), normalizeNewlines(correctResult)) {
					t.Errorf("File contents mismatch for %s.\nGot:\n%s\nWant:\n%s\n",
						tt.resultFile, currentResult, correctResult)
				}
			}
```

File: interdiff_test.go (hunk 6 above)
Comment (sofiia-tesliuk) [reply interdiff_test.go:33]: I will mark this files in next way.
For files:
`diff_f{i}_{c}.txt` -- `{c}` diff file for `file_{i}`.
`diff_f1_a.txt` -- `a` diff for `file_1.txt`.
`file_1.txt` + `diff_f1_a.txt` --> `file_1_a.txt`

For source (dir with files inside):
`diff_s{i}_{c}.txt` -- `{c}` diff file for `source_{i}`.
`diff_s1_a.txt` -- `a` diff for `source_1/`

File: interdiff_test.go (hunk 7 above)
Comment (sofiia-tesliuk) [reply interdiff_test.go:26]: Rest test cases will be corrected in another PR.

**Pull request merged**